src/main/network/relay/mod.rs

```diff
@@ -1 +1,219 @@
+use std::net::Ipv4Addr;
+use std::sync::{Arc, Weak};
+
+use atomic_refcell::AtomicRefCell;
+
+use crate::core::work::task::TaskRef;
+use crate::core::worker::Worker;
+use crate::cshadow as c;
+use crate::host::host::Host;
+use crate::network::packet::PacketStatus;
+use crate::network::relay::token_bucket::TokenBucket;
+use crate::network::{Packet, PacketDevice};
+use crate::utility::{Magic, ObjectCounter};
+use shadow_shim_helper_rs::simulation_time::SimulationTime;
+
 mod token_bucket;
+
+/// Specifies a throughput limit when forwarding packets.
+pub enum RateLimit {
+    BytesPerSecond(u64),
+    Unlimited,
+}
+
+/// Relays packets between two packet devices, optionally enforcing a bandwidth
+/// rate limit.
+pub struct Relay {
+    magic: Magic<Self>,
+    _counter: ObjectCounter,
+    // Internals in an Arc so that we can schedule tasks that refer back to it.
+    internal: Arc<AtomicRefCell<RelayInternal>>,
```

Comment:
OK, I have it mostly working except for the `safely_drop` part. Can you suggest how I should call `safely_drop`?

I can't call it inside of a `Drop` impl in `Relay` because I don't have access to the host there, which I need to get the root? So I think I have to call a new relay function from  inside of the `Drop` impl in `Host`.

But I can't figure out the correct syntax to make it compile. This fails:
```rust
    pub fn safely_drop(&self, root: &Root) {
        self.internal.safely_drop(root);
    }
```

because 
> cannot move out of `self.internal` which is behind a shared reference

Reply:
> I can't call it inside of a Drop impl in Relay because I don't have access to the host there, which I need to get the root?

You could use  `Worker::with_active_host` in the `Drop` implementation. It's a little nicer not to have to depend on that though if adding a `safely_drop` method to `Relay` is workable.

> But I can't figure out the correct syntax to make it compile. This fails:
> ...
> cannot move out of self.internal which is behind a shared reference

Yup, it's because `&self` is a shared reference. Such a method should probably take `self`, forcing the caller to invalidate its reference when calling this method. `internal` would also probably need to be wrapped in an `Option`, since `RootedRc::safely_drop` takes `self` by value, so you need to be able to dispose of that value independently from the rest of the object. e.g.:

```
pub fn safely_drop(self, root: &Root) {
    self.internal.take().unwrap().safely_drop(root);
}
```

Reply:
@sporksmith I implemented @stevenengler suggestion to place the `Arc` outside of the `Relay`, and change the function signatures to, for example:

```rust
pub fn notify(self: &Arc<Self>, host: &Host) { ... }
```

My plan was to add a new commit to switch over from `Arc` and `RefCell` to `RootedRc` and `RootedRefCell`, like this:

```rust
pub fn notify(self: &RootedRc<Self>, host: &Host) { ... }
```

But I'm getting the following compiler error:

```
`&shadow_shim_helper_rs::rootedcell::rc::RootedRc<Relay>` cannot be used as the type of `self` without the `arbitrary_self_types` feature
see issue #44874 <https://github.com/rust-lang/rust/issues/44874> for more information
consider changing to `self`, `&self`, `&mut self`, `self: Box<Self>`, `self: Rc<Self>`, `self: Arc<Self>`, or `self: Pin<P>` (where P is one of the previous types except `Self`)
```

The fix I can think of is to rename `self` and make everything non-method associated functions:

```rust
pub fn notify(rr_self: &RootedRc<Self>, host: &Host) { ... }
```

But I'm not a huge fan of calling associated functions from a design perspective.

So what's the path forward? I think this has broader implications for the rest of our code if we want to start being more consistent about which wrapper types we use (which I think we should be).

Reply:
After discussion with @sporksmith , we identified another way to handle this is to store Relays in a table in Host, set up a RelayID, and when the relay task executes look up the Relay by the RelayID. This way we wouldn't have to deal with the Weak -> Strong upgrade in Relay. (Though we think the Relay would still have to be Send because it's storing a packet.)

In any case, we thought any more changes should probably wait for a future PR, maybe when we have more consistency in how we store other objects across the codebase.